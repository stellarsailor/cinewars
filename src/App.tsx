import { useEffect } from 'react';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom'

import useList from './hooks/useList';
import Character from './pages/Character';
import Main from './pages/Main';
import AppLayout from './components/AppLayout';
import Loader from './components/Loader';

function App() {
  
  const { status, initialFetch } = useList();

  useEffect(() => {
    initialFetch(1)
  },[initialFetch])

  return (
    <Router>
      <Switch>
        {status === 'pending' && <Loader global />}
        <AppLayout>
          <Route path="/" exact>
            <Main />
          </Route>
          <Route path="/character/:id">
            <Character />
          </Route>
        </AppLayout>
      </Switch>
    </Router>
  );
}

export default App;
