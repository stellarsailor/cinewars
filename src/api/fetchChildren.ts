import { CharacterProfile } from '../types/CharacterProfile';

export async function fetchChildren(data: CharacterProfile) {
  // in iOS safari, cors issue occurs in localhost. To prevent it, replace http url to https.
  // there was cors issue, but it solves after adding meta tag <meta http-equiv="Content-Security-Policy" content="upgrade-insecure-requests">
  // it was due to SWAPI server is being served on http
  const res = await fetch(data.homeworld.replace('http:', 'https:'));
  const homeworldData = await res.json();

  const filmsPromisesArray = data.films?.map(async (url: string) =>
    fetch(url.replace('http:', 'https:')).then((res) => res.json())
  );
  // ! non-null assertion operator for -> Argument of type 'Promise<any>[] | undefined' is not assignable to parameter of type 'Iterable<any>'.
  const filmsData = await Promise.all(filmsPromisesArray!);

  const speciesPromisesArray = data.species?.map(async (url: string) =>
    fetch(url.replace('http:', 'https:')).then((res) => res.json())
  );
  const speciesData = await Promise.all(speciesPromisesArray!);

  const vehiclesPromisesArray = data.vehicles?.map(async (url: string) =>
    fetch(url.replace('http:', 'https:')).then((res) => res.json())
  );
  const vehiclesData = await Promise.all(vehiclesPromisesArray!);

  const starshipsPromisesArray = data.starships?.map(async (url: string) =>
    fetch(url.replace('http:', 'https:')).then((res) => res.json())
  );
  const starshipsData = await Promise.all(starshipsPromisesArray!);

  return {
    ...data,
    homeworldData: homeworldData,
    filmsData: filmsData,
    speciesData: speciesData,
    vehiclesData: vehiclesData,
    starshipsData: starshipsData,
  };
}
