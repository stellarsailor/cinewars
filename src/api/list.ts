export async function getList(page: number) {
  let url = `http://swapi.dev/api/people/?page=${page}`;
  try {
    const res = await fetch(url)
    const data = await res.json()
    return data
  } catch (err) {
    console.log(err)
  }
}