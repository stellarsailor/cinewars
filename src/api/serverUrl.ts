export const serverUrl = 'https://swapi.dev/api';
