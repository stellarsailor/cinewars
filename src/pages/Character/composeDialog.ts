import { CharacterProfile } from '../../types/CharacterProfile';

export function composeDialog(data: CharacterProfile) {
  let str = '<span>';

  // gender
  if (data.gender === 'male') str += 'His ';
  if (data.gender === 'female') str += 'Her ';
  if (data.gender !== 'male' && data.gender !== 'female') str += 'Its ';

  // name
  if (data.name === 'Darth Vader') {
    return (str = 'Lord Vader, What more can I say?');
  }
  str +=
    'name is <span style="font-weight:bold;">' + data.name + '</span> and has ';

  // hair_color
  if (data.hair_color === 'none' || data.hair_color === 'n/a')
    str += 'no hair and';
  else {
    let splitHair = data.hair_color.split(', ');
    //exceptions: blond, auburn, black(same background color)
    let mainHair = splitHair[0].replace(`"`, '');
    if (mainHair === 'blond') mainHair = '#faf0be';
    if (mainHair === 'auburn') mainHair = '#A52A2A';
    if (mainHair === 'black') mainHair = '#4f4f4f';
    str += `<span style="font-weight:bold; color:${mainHair};"> ${data.hair_color} hair</span> and `;
  }

  // eye_color
  let mainEyeColor = data.eye_color;
  //exceptions: blue-gray, blue, hazel, black(same background color)
  if (mainEyeColor === 'blue-gray') mainEyeColor = '#6699cc';
  if (mainEyeColor === 'blue') mainEyeColor = '#a1caf1';
  if (mainEyeColor === 'hazel') mainEyeColor = '#a5c6bb';
  if (mainEyeColor === 'black') mainEyeColor = '#4f4f4f';
  str += `<span style="font-weight:bold; color:${mainEyeColor};"> ${data.eye_color} eyes</span>. `;

  // confirmation
  if (data.gender === 'male') str += 'Is he the one you are looking for?';
  if (data.gender === 'female') str += 'Is she the one you are looking for? ';
  if (data.gender !== 'male' && data.gender !== 'female')
    str += `I guess, these aren't the Droids you're looking for.`;

  str += '</span>';
  return str;
}
