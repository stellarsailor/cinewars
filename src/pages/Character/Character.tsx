import { useEffect, useState } from "react"
import { useParams } from 'react-router-dom'
import styled from "styled-components"
import { fetchChildren } from "../../api/fetchChildren"
import { serverUrl } from "../../api/serverUrl"
import CharacterDetail from "../../components/CharacterDetail"
import DialogBox from "../../components/DialogBox"
import Loader from "../../components/Loader"
import useDialog from "../../hooks/useDialog"
import { CharacterProfile } from "../../types/CharacterProfile"
import { composeDialog } from "./composeDialog"

function Character() {

  const params: any = useParams()
  const { dialog, onSetDialog } = useDialog()

  const [ character, setCharacter ] = useState<null | CharacterProfile>(null)
  const [ loading, setLoading ] = useState(false)

  useEffect(() => {
    const initialFetch = async () => {
      if(params === undefined) return
      let url = `${serverUrl}/people/${params.id}`
      setLoading(true)
      try {
        const res = await fetch(url)
        const data = await res.json()
        const characterData = await fetchChildren(data)

        onSetDialog(composeDialog(data))
        setCharacter(characterData)
        setLoading(false)
      } catch (err) {
        console.log(err)
      }
    }
    initialFetch()
  },[params, onSetDialog])

  if(loading) return <Loader global />

  return (
    <Container>
      <CharacterDetail
        id={params.id}
        character={character}
      />
      <DialogBox
        content={dialog.text}
      />
    </Container>
  )
}

const Container = styled.div`
  z-index: 10;
`

export default Character