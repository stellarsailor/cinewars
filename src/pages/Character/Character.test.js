import { 
  render, 
  fireEvent, 
  waitFor 
} from "@testing-library/react";
import { 
  createStore, 
  applyMiddleware 
} from 'redux'
import { Provider } from 'react-redux'
import Thunk from 'redux-thunk';
import rootReducer from '../../modules';
import App from '../../App'
import Character from "./Character";
import { serverUrl } from "../../api/serverUrl";
import { createMemoryHistory } from 'history'
import { Router, Route, Switch } from 'react-router-dom'
import axios from 'axios';
// to resolve cors problem (https://github.com/axios/axios/issues/1754)
axios.defaults.adapter = require('axios/lib/adapters/http');

const store = createStore(rootReducer, applyMiddleware(Thunk))

describe('<Character />', () => {
  it('matches snapshot', () => {
    const history = createMemoryHistory()
    const utils = render(
      <Provider store={store}>
        <Router history={history}>
          <Character />
        </Router>
      </Provider>
    );
    expect(utils.container).toMatchSnapshot();
  });

  it('shows the loader indicator properly', () => {
    const history = createMemoryHistory()
    const route = '/character/1'
    history.push(route)

    const utils = render(
      <Provider store={store}>
        <Router history={history}>
          <Character />
        </Router>
      </Provider>
    );
    
    utils.getByAltText("loader indicator")
  });

  // This should test one character's info properly, but it cannot fetch data from API due to "Error: Cross origin http://localhost forbidden". It can be solved by changing  fetch() to axios() like line 20 here, but the instruction said "Use the Fetch API". I left the function with fetch().

  // it('should fetch data properly with fetch() function', async () => {
  //   const history = createMemoryHistory()
  //   const route = '/character/1'
  //   history.push(route)

  //   const utils = render(<Provider store={store}><Router history={history}><Character /></Router></Provider>);

  //   let url = `${serverUrl}/people/1`
  //   const res = await axios(url)
  //   const data = res.data

  //   await waitFor(() => {
  //     utils.getByText(data.name)
  //     utils.getByText(data.birth_year)
  //     utils.getByText(data.height)
  //     utils.getByText(data.mass)
  //   })
  // })

})