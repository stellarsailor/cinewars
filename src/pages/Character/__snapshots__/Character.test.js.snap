// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Character /> matches snapshot 1`] = `
<div>
  <div
    class="sc-hBEYos fvYwXF"
  >
    <img
      alt="loader indicator"
      height="40px"
      src="/images/loader.png"
      width="40px"
    />
  </div>
</div>
`;
